packages/react/src/ReactInlineContentSpec.tsx

```diff
@@ -0,0 +1,152 @@
+import { FC } from "react";
+import { flushSync } from "react-dom";
+import { createRoot } from "react-dom/client";
+
+// this file is mostly analogoues to `customBlocks.ts`, but for React blocks
+
+// extend BlockConfig but use a React render function
+export type ReactInlineContentImplementation<
+  T extends InlineContentConfig,
+  // I extends InlineContentSchema,
+  S extends StyleSchema
+> = {
+  render: FC<{
+    inlineContent: InlineContentFromConfig<T, S>;
+    contentRef: (node: HTMLElement | null) => void;
+  }>;
+  // TODO?
+  // toExternalHTML?: FC<{
+  //   block: BlockFromConfig<T, I, S>;
+  //   editor: BlockNoteEditor<BlockSchemaWithBlock<T["type"], T>, I, S>;
+  // }>;
+};
+
+// A function to create custom block for API consumers
+// we want to hide the tiptap node from API consumers and provide a simpler API surface instead
+export function createReactInlineContentSpec<
+  T extends InlineContentConfig,
+  // I extends InlineContentSchema,
+  S extends StyleSchema
+>(
+  inlineContentConfig: T,
+  inlineContentImplementation: ReactInlineContentImplementation<T, S>
+) {
+  const node = createStronglyTypedTiptapNode({
+    name: inlineContentConfig.type as T["type"],
+    inline: true,
+    group: "inline",
+    content: (inlineContentConfig.content === "styled"
+      ? "inline*"
+      : "") as T["content"] extends "styled" ? "inline*" : "",
+
+    addAttributes() {
+      return propsToAttributes(inlineContentConfig.propSchema);
+    },
+
+    // parseHTML() {
+    //   return parse(blockConfig);
+    // },
+
+    renderHTML({ node }) {
+      const editor = this.options.editor;
+
+      const ic = nodeToCustomInlineContent(
+        node,
+        editor.inlineContentSchema,
+        editor.styleSchema
+      ) as any as InlineContentFromConfig<T, S>; // TODO: fix cast
+
+      const Content = inlineContentImplementation.render;
+
+      let contentDOM: HTMLElement | undefined;
+      const div = document.createElement("div");
+      const root = createRoot(div);
+      flushSync(() => {
+        root.render(
+          <Content
+            inlineContent={ic}
+            contentRef={(el) => (contentDOM = el || undefined)}
+          />
+        );
+      });
+
+      if (!div.childElementCount) {
+        // TODO
+        console.warn("ReactInlineContentSpec: renderHTML() failed");
+        return {
+          dom: document.createElement("span"),
+        };
+      }
+
+      // clone so we can unmount the react root
+      contentDOM?.setAttribute("data-tmp-find", "true");
+      const cloneRoot = div.cloneNode(true) as HTMLElement;
+      const dom = cloneRoot.firstElementChild! as HTMLElement;
+      const contentDOMClone = cloneRoot.querySelector(
+        "[data-tmp-find]"
+      ) as HTMLElement | null;
+      contentDOMClone?.removeAttribute("data-tmp-find");
+
+      root.unmount();
+
+      return {
+        dom,
+        contentDOM: contentDOMClone || undefined,
+      };
+    },
+
+    // TODO: needed?
```

Comment:
I think it's a nice to have but I don't think there are many (if any) use cases that would take advantage of it. The custom inline content type would either have to reference its parent block or attach listeners to elements other than itself. I guess the only concern is that a `ReactNodeViewRenderer` is how TipTap recommends rendering a React component for your node, so might have to look inside their code to see if there's anything we're missing.